### ISSUE TITLE
Unchecked exception audit

### ISSUE THREAD

**Issue by ricellis** [cloudant/sync-android#280]
There are far too many `RuntimeException` and other unchecked exceptions.
We should go through and remove these where possible to make error handling easier.
We should also pay particular attention to the package com.cloudant.sync.query which could use some exception handling improvements.

**Comment by tomblench** [comment]
Note also that some SQL errors don't manifest themselves in obvious ways because the user has to look at the log to see what exactly happened. This is due to the way that `insert` and `insertWithOnConflict` do not throw exceptions (this is by design - it is compatible with the Android SQL API).

See below for an example log. The user can't programmatically retrieve the underlying error because `InsertRevisionCallable.call` throws this exception:

```
throw new IllegalStateException("Unknown error inserting new revision, please check log");
```

```
SEVERE: Error inserting to: revs, parent=25 current=false deleted=true revid=26-9f17ab457a9f4d38e0b8fa02e9af6db9 available=true json=[B@64accba4 doc_id=1, 
com.almworks.sqlite4java.SQLiteException: [2067] DB[1] step() [INSERT  INTO "revs"(parent,current,deleted,revid,available,json,doc_id) VALUES (?,?,?,?,?,?,?)]DB[1][C] [UNIQUE constraint failed: revs.doc_id, revs.revid]
	at com.almworks.sqlite4java.SQLiteConnection.throwResult(SQLiteConnection.java:1436)
	at com.almworks.sqlite4java.SQLiteConnection$BaseController.throwResult(SQLiteConnection.java:1689)
	at com.almworks.sqlite4java.SQLiteStatement.stepResult(SQLiteStatement.java:1402)
	at com.almworks.sqlite4java.SQLiteStatement.step(SQLiteStatement.java:301)
	at com.cloudant.sync.sqlite.sqlite4java.SQLiteWrapper.executeSQLStatement(SQLiteWrapper.java:331)
	at com.cloudant.sync.sqlite.sqlite4java.SQLiteWrapper.insertWithOnConflict(SQLiteWrapper.java:314)
	at com.cloudant.sync.sqlite.sqlite4java.SQLiteWrapper.insert(SQLiteWrapper.java:325)
	at com.cloudant.sync.datastore.callables.InsertRevisionCallable.call(InsertRevisionCallable.java:67)
	at com.cloudant.sync.datastore.DatastoreImpl.insertDocumentHistoryIntoExistingTree(DatastoreImpl.java:1377)
	at com.cloudant.sync.datastore.DatastoreImpl.doForceInsertExistingDocumentWithHistory(DatastoreImpl.java:1337)
	at com.cloudant.sync.datastore.DatastoreImpl.access$800(DatastoreImpl.java:78)
	at com.cloudant.sync.datastore.DatastoreImpl$14.call(DatastoreImpl.java:1170)
	at com.cloudant.sync.sqlite.SQLQueueCallable.call(SQLQueueCallable.java:36)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Aug 23, 2016 3:11:36 PM com.cloudant.sync.replication.PullStrategy run
SEVERE: Batch 1 ended with error:
java.util.concurrent.ExecutionException: com.cloudant.sync.datastore.DocumentException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Unknown error inserting new revision, please check log
	at com.cloudant.sync.replication.PullStrategy.processOneChangesBatch(PullStrategy.java:409)
	at com.cloudant.sync.replication.PullStrategy.replicate(PullStrategy.java:253)
	at com.cloudant.sync.replication.PullStrategy.run(PullStrategy.java:178)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.cloudant.sync.datastore.DocumentException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Unknown error inserting new revision, please check log
	at com.cloudant.sync.datastore.DatastoreImpl.forceInsert(DatastoreImpl.java:1266)
	at com.cloudant.sync.replication.DatastoreWrapper.bulkInsert(DatastoreWrapper.java:107)
	at com.cloudant.sync.replication.PullStrategy.processOneChangesBatch(PullStrategy.java:407)
	... 3 more
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Unknown error inserting new revision, please check log
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.cloudant.sync.datastore.DatastoreImpl.forceInsert(DatastoreImpl.java:1255)
	... 5 more
Caused by: java.lang.IllegalStateException: Unknown error inserting new revision, please check log
	at com.cloudant.sync.datastore.callables.InsertRevisionCallable.call(InsertRevisionCallable.java:69)
	at com.cloudant.sync.datastore.DatastoreImpl.insertDocumentHistoryIntoExistingTree(DatastoreImpl.java:1377)
	at com.cloudant.sync.datastore.DatastoreImpl.doForceInsertExistingDocumentWithHistory(DatastoreImpl.java:1337)
	at com.cloudant.sync.datastore.DatastoreImpl.access$800(DatastoreImpl.java:78)
	at com.cloudant.sync.datastore.DatastoreImpl$14.call(DatastoreImpl.java:1170)
	at com.cloudant.sync.sqlite.SQLQueueCallable.call(SQLQueueCallable.java:36)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	... 1 more

**Comment by tomblench** [comment]
This should also be the place to address general `DatastoreImpl` exception design issues. We have too many public API methods which don't have any `throws` clauses on their signatures and just `return null` when something goes wrong.

**Comment by ricellis** [comment]
The `ReplicatorBuilder` throws `RuntimeException`s for bad `URL`s. Whilst this would indicate an application error, throwing an unchecked exception does not seem like the correct choice.

**Comment by tomblench** [comment]
The query code on the 2.0 branch has improved but there are still internal classes like `QueryExecutor` which return null in lots of places where they should throw.

**Comment by tomblench** [comment]
@ricellis now that #420 and #425 are merged can we consider this done?

**Comment by ricellis** [comment]
Nope, whilst those provide an excellent start there are still places that need checking before a 2.0. The `ReplicatorBuilder` cited above is an example, but there may be others too.

**Comment by tomblench** [comment]
#429 addresses the issues in `ReplicatorBuilder` but there are still 41 instances of `RuntimeException` being thrown (not including sub-classes). So these will need to be audited.

**Comment by rhyshort** [comment]
Runtime Exceptions Remaining (on feature branch)
----------

> All New RuntimeException creations in production code from commit 14a827b99fa19670980188b83a242b7f1ac509d9 on branch feature-284-refactor-datastore-indexmanager

1. com.cloudant.sync.event.EventBus#post(Object)
   doesn't get wrapped, we should probably swallow error and log message.


1. com.cloudant.sync.internal.documentstore.DatabaseImpl#createDocumentBody()
	gets wrapped as ExecutionException and then DocumentStoreException.

1. LibraryVersion, removed on master need to bring commit across.

1. ForceInsert for interrupted exception, wrapped as ExecutionException in PullStrategy

1. getFullRevisionFromCurrentCursor throws runtime exception for unexpected column type for parent. Will get wrapped in ExecutionException.

1.	getPossibleAncestorRevisionIDs throws runtime will get wrapped in ExecutionException in PullStrategy

1. MultipartAttachment#setup thrown when UTF-8 encoding is missing no action required.

1. EncryptedAttachmentInputStream many occurrences failure to init crypto engine eventually wrapped by execution in replicator, execution on SQLDatabase Queue.

1. EncryptedAttachmentOutputstream many instances, in ctr, wrapped into execution Exception by SQLDatabaseQueue and replicator.

1. encodeId throws runtime when UTF-8 Is missing, no action required.

1. encodeQueryParameter UTF-8 missing.

1. uriFor turning string into URI, thrown when URISyntaxException is thrown. Eventually caught by replicator, not wrapped. in some cases.

1. DoucmentRevs#setOthers thrown when key begins with `_`, thrown to be defensive, will cause problems if new fields added to documents.

1. JSONHelper thrown in many places seems to be eventually wrapped / passed back to the user via the replicator.

1. QueryImpl listIndexes throws when listing index fails. Leeks out to user without wrapping. 

1. bulkCreateSerialisedDocs, thrown when COUCH API returns docs for insert, wrapped by replicator.

1. getRevisions, thrown when missing is returned when using `open_revs` shouldn't be a problem unless the rev doesn't exist or has never existed.

1. GetRevisionTaskIterator#next throws, when interrupted and executionException encountered, nothing better can be done here restricted by the iterator interface. Hard to determine user exposure.

1. makePlaceholders thrown if len is < 1. Should be an illegal Arg. eventually wrapped by SQLQueue.

1. toPrettyJSSon exposed via error info in replicator. (JSONUtils need to be merged with JSONHelper.)

1. ReplicatorBuilder been fixed.

**Comment by ricellis** [comment]
@rhyshort comments added inline in your comment.

**Comment by tomblench** [comment]
12. "if a user provided URI string can cause the runtime exception..." This shouldn't be possible because the only place the user provides a URI is in the ReplicatorBuilder and it has to be valid (because it's not possible to construct an invalid URI)

**Comment by rhyshort** [comment]
List of places where action is required for RuntimeExceptions 



1. com.cloudant.sync.event.EventBus#post(Object)
   doesn't get wrapped, we should probably swallow error and log message.
    - we should probably check for this at `register` time and throw an `IllegalArgumentException`, then the `RuntimeException` at `post` should never happen
        - @rhyshort working on making the check at `register` time. PR #441 

1. ~~LibraryVersion, removed on master need to bring commit across.~~

1.	getPossibleAncestorRevisionIDs throws runtime will get wrapped in ExecutionException in PullStrategy
    - This should probably become something checked

1. JSONHelper thrown in many places seems to be eventually wrapped / passed back to the user via the replicator.
    - Merge with `JSONUtils` ensure all deserialization derived exceptions will go through to `ReplicationErrored`. If there are other callers will need to investigate case by case. @rhyshort is working on.
        - #438 merged JSONUtils and JSONHelper, still need to make sure exceptions don't leak however.

1. QueryImpl listIndexes throws when listing index fails. Leaks out to user without wrapping.
    - Being fixed #431 


1. makePlaceholders thrown if len is < 1. Should be an illegal Arg. eventually wrapped by SQLQueue.
    - This should become an `IllegalArgumentException`, but we also need to check that we don't call with an empty list, so we should add defensive checks further up the stack.

1. ~~toPrettyJSon exposed via error info in replicator. (JSONUtils need to be merged with JSONHelper.)~~
    - As per `JSONHelper`
    - JSONHelper merged into JSONUtils in #438 

1. ~~ReplicatorBuilder will be fixed through later pr.~~
   - #435
